contracts/test-utils/src/unit_test_helper.ts

```diff
@@ -0,0 +1,68 @@
+import { RevertError } from '@0x/utils';
+
+import { TransactionReceiptWithDecodedLogs } from 'ethereum-types';
+
+import { expect } from './chai_setup';
+
+type AsyncFunction<TArgs extends any[], TReturn> = (...args: TArgs) => Promise<TReturn>;
+
+interface TransactionReturnData<TCallAsyncResult> {
+    result: TCallAsyncResult;
+    receipt: TransactionReceiptWithDecodedLogs;
+}
+
+interface ContractWrapperFunction<
+    TCallAsyncArgs extends any[],
+    TAwaitTransactionSuccessAsyncArgs extends any[],
+    TCallAsyncResult
+> {
+    callAsync: AsyncFunction<TCallAsyncArgs, TCallAsyncResult>;
+    awaitTransactionSuccessAsync?: AsyncFunction<TAwaitTransactionSuccessAsyncArgs, TransactionReceiptWithDecodedLogs>;
+}
+
+export type TransactionExpectation<TAwaitTransactionSuccessAsyncArgs extends any[], TCallAsyncResult> = (
+    tx: Partial<TransactionReturnData<TCallAsyncResult>>,
+    // tslint:disable-next-line: trailing-comma
+    ...args: TAwaitTransactionSuccessAsyncArgs
+) => Promise<void> | void;
+
+export class BaseUnitTestHelper<TContract> {
+    protected static async _verifyContractMethodExpectationsAsync<
+        TCallAsyncArgs extends any[],
+        TAwaitTransactionSuccessAsyncArgs extends any[],
+        TCallAsyncResult
+    >(
+        contractFunction: ContractWrapperFunction<TCallAsyncArgs, TAwaitTransactionSuccessAsyncArgs, TCallAsyncResult>,
+        expectations: Array<TransactionExpectation<TAwaitTransactionSuccessAsyncArgs, TCallAsyncResult>> = [],
```

Comment:
I wonder if it just makes sense to have this be a single `TransactionExpectation` callback, since it can always wrap an array of tests, and it can be more flexible about ordering or using intermediate results. But then I guess this function would pretty much just boil down to `TransactionHelper.getResultAndReceipt()` for mutator functions... but those are the only functions where we would really need this abstraction anyway? But if we're gonna just pass a single callback, then we might as well just return `result` and `receipt` and leave the responsibility to the caller to do the assertions on them... which is exactly what `TransactionHelper` does. ┐('～`;)┌

Reply:
You're totally right, `TransactionHelper` does like 90% the same things as this class. 
But to answer your main question of why we need this when we have `TransactionHelper`, the tl;dr is that we need a layer of abstraction between the functionality provided by `TransactionHelper` (low level) and our unit tests (should be high level, but currently do a lot of heavy lifting) in order to avoid the rampant code duplication. You can see what having a middle layer of abstraction does for our tests [here](https://github.com/0xProject/0x-monorepo/pull/2143/files). 

Note that `BaseUnitTestHelper` is _not_ that middle layer, but I believe it serves as a better base for such a layer than `TransactionHelper` because of the modularity provided by the `expectations` abstraction. We'll be able to share a single `expectation` callback between multiple tests, so the idea is to have these `TransactionExpectation`s in the derived unit test helper classes along with other utility functions so that the actual unit test files are as clear and concise as possible.